sql/src/main/java/io/crate/executor/transport/ExecutionNodesTask.java

```diff
@@ -353,29 +400,31 @@ public int buckedIdx() {
                         return bucketIdx;
                     }
                 });
-            } else {
-                pageDownstreamContext.failure(bucketIdx, new IllegalStateException("expected directResponse but didn't get one"));
             }
         }
 
         @Override
         public void onFailure(Throwable e) {
-            pageDownstreamContext.failure(bucketIdx, e);
+            for (PageDownstreamContext pageDownstreamContext : pageDownstreamContexts) {
+                pageDownstreamContext.failure(bucketIdx, e);
+            }
         }
     }
 
     private static class FailureOnlyResponseListener implements ActionListener<JobResponse> {
 
-        private final SettableFuture<TaskResult> result;
+        private final List<SettableFuture<TaskResult>> results;
 
-        public FailureOnlyResponseListener(SettableFuture<TaskResult> result) {
-            this.result = result;
+        public FailureOnlyResponseListener(List<SettableFuture<TaskResult>> results) {
+            this.results = results;
         }
 
         @Override
         public void onResponse(JobResponse jobResponse) {
-            if (jobResponse.directResponse().isPresent()) {
-                result.setException(new IllegalStateException("Got a directResponse but didn't expect one"));
+            if (jobResponse.directResponse().size() > 0) {
+                for (SettableFuture<TaskResult> result : results) {
```

Comment:
is it really required to loop over the results? if size() > 0  you could set the exception directly.